designs/2021-suppression-support/README.md

```diff
@@ -0,0 +1,459 @@
+            }
+          }
+        }
+      ],
+      "suppressions": [
+        {
+          "kind": "inSource",
+          "justification": "Sample justification message."
+        }
+      ],
+      "ruleId": "no-undef",
+      "ruleIndex": 0
+    }
+  ]
+}
+```
+
+Input:
+
+```
+.\node_modules\.bin\eslint.cmd 1.js -f json -o 1.json
+```
+
+Output:
+
+```json
+JSON format
+[
+  {
+    "filePath":"C:\\Users\\yiweiding\\Documents\\projects\\test\\1.js",
+    "messages":[
+      { ... another record ... }
+    ],
+    "suppressedMessages":[
+      {
+        "ruleId":"no-undef",
+        "severity":2,
+        "message":"'a' is not defined.",
+        "line":1,
+        "column":1,
+        "nodeType":"Identifier",
+        "suppressions":[
+          {
+            "kind":"directive",
+            "justification":"Sample justification message."
+          }
+        ],
+        "messageId":"undef",
+        "endLine":1,
+        "endColumn":2
+      }
+    ],
+    "errorCount":0,
+    "warningCount":0,
+    "fixableErrorCount":0,
+    "fixableWarningCount":0,
+    "usedDeprecatedRules":[]
+  }
+]
+```
+
+### Parse description in comments as `Justification`
+
+We propose not to ignore the text after two or more dashes in comments and considered it as the justification of a suppression. The modification should be in `getDirectiveComments`.
+
+### Add `SuppressedLintMessage` inheriting from `LintMessage`
+
+According to this [comment](https://github.com/eslint/rfcs/pull/82#discussion_r733584774), we are going to have a `suppressedMessages` distinguished from `messages` in the output of ESLint. Items in `suppressedMessages` should have the property `suppressions`, a list of suppressions (`{kind: "directive", justification: "Justification message example."}`).
+
+### Reserve directive suppression
+
+In the step of apply directive suppression, ESLint will not remove the problems that should be disabled by directive comments. These problems will be reserved and added `{kind: "directive", justification: "Justification message example."}`.
+
+NOTE that the kind `directive` should be converted to `inSource` for SARIF according to [kind property](https://docs.oasis-open.org/sarif/sarif/v2.1.0/os/sarif-v2.1.0-os.html#_Toc34317735).
+
+### Modify `CLIEngine` and related formatters
+
+As mentioned, we are not going to change the current behavior of ESLint. Without `--no-inline-config`, ESLint should not report errors/warnings in CLI. In `CLIEngine`, when `suppressions` of a message is not null or empty (which means this message should be suppressed and not be reported), this message would be skipped or ignored. Additionally, the related formatters, such as stylish, should also be modified.
+
+## Documentation
+
+Add entries for the new added/modified types, `SuppressedLintMessage` and `LintResult`, to _Developer Guide – Node.js API_.
+
+## Drawbacks
+
+Related formatters, such as stylish, should be modified to keep the current behavior.
```

Comment:
Now that the proposal is to always return suppressed messages in a separate property, there is no need to update existing formatters?